_posts/2025-09-01-semantic-router.md

```diff
@@ -0,0 +1,112 @@
+* When a query indicates commercial intent (e.g., booking flights or finding legal services), it is routed to high-compute models or agent services directly integrated into transaction flows.
+
+In these cases, companies like OpenAI can participate in the value chain by taking a commission on completed transactions — transforming free usage from a cost center into a monetizable entry point.
+
+Other companies are adopting similar strategies:
+
+* **Anthropic Claude 3.7/4**: Combines "fast thinking" and "slow thinking" with user-controlled toggles.
+* **Google Gemini 2.5**: Introduces a *thinking budget*, giving enterprises fine-grained control over inference costs.
+* **Alibaba Qwen3**: Explores instruction-based switching between reasoning and non-reasoning modes.
+* **DeepSeek v3.1**: Implements a "single-model dual-mode" design, merging dialogue and reasoning.
+
+In short: the industry is entering an era where **no token should be wasted**.
+
+---
+
+## Recent Research: vLLM Semantic Router
+
+Amid this shift toward hybrid reasoning, we focus on the **open-source inference engine vLLM**.
+
+While vLLM has become the de facto standard for deploying large models, it lacks fine-grained, semantic-level control — the ability to make routing decisions based on meaning rather than query type alone. Developers are often forced to either enable full inference (wasting computation) or disable it entirely (sacrificing accuracy).
+
+To address this, we propose the **vLLM Semantic Router**, which brings GPT-5-style "smart routing" to the open-source ecosystem.
+
+![](/assets/figures/semantic-router/architecture.png)
+
+### Architecture Design
+
+1. **Semantic Classification**: Uses a **ModernBERT** fine-tuned intent classifier to determine whether a query requires inference.
+2. **Smart Routing**:
+
+   * Simple queries → Fast inference mode.
+   * Complex queries → Chain-of-Thought for accurate reasoning.
+3. **High-Performance Engine**: Built with Rust and the Hugging Face Candle framework, enabling high concurrency and zero-copy efficiency.
+4. **Cloud-Native Integration**: Seamlessly integrates with Kubernetes and API Gateways via the Envoy `ext_proc` plugin for enterprise deployments.
+
+Experimental results show:
+
+* **Accuracy**: +10.2%
+* **Latency**: –47.1%
+* **Token Consumption**: –48.5%
+
+In knowledge-intensive areas such as business and economics, accuracy improvements can exceed **20%**.
+
+---
+
+## Project Background
+
+The Semantic Router is not the isolated result of a single paper but a collaborative outcome of sustained community contributions:
+
+* Originally proposed by **Dr. Chen Huamin**, Distinguished Engineer at **Red Hat**, in early **2025** across multiple open-source communities.
+* Iterated and further developed by **Xunzhuo Liu** at **Tencent**, later contributed to the vLLM community.
+* **Dr. Wang Chen** from **IBM Research** and **Dr. Chen Huamin** will present the project at **KubeCon North America 2025**.
```

Comment:
Is it possible to add some links to these people?

Reply:
good point!

For links:
- Huamin Chen: Google Scholar (https://scholar.google.com/citations?user=6pjd6sIAAAAJ&hl=en) or linkedin (https://www.linkedin.com/in/huaminchen/)
- Chen Wang: DBLP (https://dblp.org/pid/82/4206-39.html), Google Scholar (https://scholar.google.com/citations?user=JL6iWLgAAAAJ&hl=en) or linkedin (linkedin.com/in/chenw615) @wangchen615

@Xunzhuo can you help update? Thanks

Reply:
Yep, i will add your linkedin link here

Reply:
Done.